Add Invenio RDM repository integration

davelopez opened the pull request:
First attempt at integrating [InvenioRDM](https://inveniosoftware.org/products/rdm/) into Galaxy.

The idea is to be able to import files directly from InvenioRDM repositories into Galaxy as well as publish records containing artifacts (Histories, datasets, etc.) from Galaxy to InvenioRDM.

Please note that [Zenodo is also transitioning to ZenodoRDM](https://blog.zenodo.org/2022/12/07/2022-12-07-zenodo-on-inveniordm/) therefore this integration will be useful for both once they have migrated their systems.

## Invenio RDM as a Remote File Source Plugin

The first low-hanging fruit is to integrate it as a file source plugin. This will list every record as a directory and every file inside a record as a remote file available for import.

## API changes

Since the _Remote File Source Plugin_ API is not a perfect fit for these kinds o repositories, especially the publication part, it requires some changes to the current API. I tried to keep the changes to a minimum and as generic as possible so other plugins in the future can reuse them:

- `writeable` optional flag parameter in the `GET /api/remote_files` endpoint to express the intent of writing to the remote file source when listing the files. This is required to be able to list only those records that can still upload files (are not yet published). I think this could be useful for other plugins as well, maybe to differentiate between read-only directories and writable ones.

- New endpoint `POST /api/remote_files` to create a new record (or directory) in the remote file source. This is required to be able to create a _draft record_ from Galaxy before trying to upload files to it. Other plugins can implement it as a way of creating new directories in the remote file source, by default it will raise a `NotImplementedError`.

- `include_kind` and `exclude_kind` optional parameters in the `GET /api/remote_files/plugins` endpoint to filter out plugins based on their "kind". The new `PluginKind` class variable defines the "kind" (or category) that groups similar plugins.

### Current known limitations:

- The amount of metadata entered when creating a new record is very limited, mostly the `title` and the `username` (or _Anonymous Galaxy User_ if no name is specified in the user's settings). This is to keep the Remote File Sources API as generic as possible. However, the metadata can always be edited directly in the InvenioRDM UI if required.
- There is no support for paginating the listing of records from the file source while the Invenio API forces pagination and limits the number of items returned. This might be supported in the future, for now limiting the max number of records to 100.
- Only public records can be imported for now. To be able to import private or protected records we need to pass the `user_context` when fetching the files and there is no user context available AFAIK at that point, but maybe that can be fixed.
- The `export_remote` tool is supported, but there are some limitations on the valid paths that can be selected:
  - Only records that are in draft mode are valid, so selecting the root of the repository or creating additional "directories" in the record will make the tool fail. Fortunately, the error message in the logs should be informative enough.

## For Users <a href="#for-users" id="for-users"/>

As a user, you can now import files from Invenio RDM repositories into Galaxy as well as publish records containing artifacts (Histories, datasets, etc.) from Galaxy to Invenio RDM.

### Importing files from Invenio RDM

If you just want to import from published records in Invenio RDM, you can do so by selecting the Invenio RDM repository as a remote file source in the _Upload_ tool and then selecting the records or individual files you want to import to your history. Just like you would do with any other remote file source in Galaxy.

https://github.com/galaxyproject/galaxy/assets/46503462/2a49403f-8c9e-47a7-b083-c38e31339604

By default, only published records will be listed. If you want to import from your own protected records, you will need to set your Invenio RDM API personal token in your Galaxy account settings. See the next section for more details.

### Exporting to Invenio RDM

#### Setting your Invenio RDM API personal token

First, you will need to set some preferences in your Galaxy account. Go to `User > Preferences > Manage Information` and set your Invenio RDM API personal token. You can also set some other options there, like your `Public Name` or whether you want to keep your records only accessible to you.

Galaxy will **never publish your records** (i.e. take them out of draft) as this must be done consciously from the Invenio RDM UI.

![InvenioUserSettings](https://github.com/galaxyproject/galaxy/assets/46503462/ba49924f-7c16-458d-9934-33f2dd5e2bf0)


You can create a new personal token in your Invenio RDM account settings if you don't have one.

![InvenioPersonalToken](https://github.com/galaxyproject/galaxy/assets/46503462/73c1ad11-af0e-43d3-8c2a-37a3f01eb8f0)


#### Exporting your Histories

Once you have set your personal token, you can publish your histories by selecting the Invenio RDM repository as a remote file source in the _Export History to File_ menu option. You will be able to select an existing draft record or create a new one and the files will be uploaded to it.

Once exported, you will be able to import back the history snapshot from the published record at any time.


https://github.com/galaxyproject/galaxy/assets/46503462/8d9cf938-fdd2-48e9-8e34-ed2f910598d7



#### Archiving your Histories

When you are completely done and happy with your history, you can archive it to take it out of your active histories and even free up some disk quota. You can do so by selecting the _Archive History_ menu option. The process will be similar to exporting the history, but your history will be purged from Galaxy once the process is completed.

You will be able to import back the history as a copy from the published record at any time.

In the video below, we will first create the draft record directly from Invenio so we can provide as much metadata as we want.

https://github.com/galaxyproject/galaxy/assets/46503462/26e641c2-1c70-430a-85bb-e017b0464670



#### Exporting your Datasets

You can also publish individual datasets to Invenio RDM by selecting the Invenio RDM repository as a remote file source using the _Export Datasets_ tool. Since the tool interface is a bit limited, you will need to create your draft record manually in advance, and then you can select it when exporting the datasets.

In the video below, we will first create the draft record directly from Invenio as we must select an existing record from the tool.

https://github.com/galaxyproject/galaxy/assets/46503462/b3df659b-57c4-4103-bd3f-4e9ade9bb829



## For Reviewers

This PR includes a bunch of UI modernizations (Composition API + Typescript) for the components that have been modified. To simplify the review, those standalone changes have been extracted to individual PRs:

- [x] #16414
- [x] #16391
- [x] #16420

### Manual Testing Requirements <a href="#requirements" id="requirements"/>

To test the integration you need an **Invenio RDM backend**, ideally, a sandbox environment where you are not really publishing stuff to the world.

- **Create your local instance** by following the [Quick Start documentation](https://inveniordm.docs.cern.ch/install/) (using the `Containerized application` is recommended). 
- You would need to create a user and activate it in your Invenio instance, the [documentation also covers this](https://inveniordm.docs.cern.ch/customize/vocabularies/users/#users).
- Then add to your Galaxy `file_sources_conf.yml` a new entry with something like this:
  ```yaml
  - type: inveniordm
    id: invenio
    doc: Invenio RDM The turn-key research data management repository
    label: Invenio RDM (Local Test)
    url: https://127.0.0.1
    token: ${user.preferences['invenio|token']}
    public_name: ${user.preferences['invenio|public_name']}
    writable: true
  ```
- Also add a new entry in your `user_preferences_extra_conf.yml` with the following:
  ```yaml
  invenio:
      description: Your Invenio RDM Account
      inputs:
          - name: token
            label: Personal Token to publish records to Invenio RDM
            type: secret
            store: vault # Requires setting up vault_config_file in your galaxy.yml
            required: False
          - name: public_name
            label: Public name to publish records (formatted as "Lastname, Firstname")
            type: text
            required: False
          - name: public_records
            label: Whether to make new draft records publicly accessible (Yes) or restricted (No).
            type: boolean
            required: False
  ```
- If you don't have a **Vault** configured in your local instance you should set it up to store the personal token as a secret. Check the [documentation](https://github.com/galaxyproject/galaxy/blob/dev/doc/source/admin/special_topics/vault.md).
- The [new task-based export UI that tracks export records](https://github.com/galaxyproject/galaxy/pull/14839) requires having Celery Task (with Redis backend) setup in your instance. So you will need something like this in your `galaxy.yml`:
  ```yml
     # For details, see Celery documentation at
      # https://docs.celeryq.dev/en/stable/userguide/configuration.html.
      celery_conf:
          result_backend: redis://127.0.0.1:6379/0
      #  task_routes:
      #    galaxy.fetch_data: galaxy.external
      #    galaxy.set_job_metadata: galaxy.external
  
      # Offload long-running tasks to a Celery task queue. Activate this
      # only if you have setup a Celery worker for Galaxy. For details, see
      # https://docs.galaxyproject.org/en/master/admin/production.html
      enable_celery_tasks: true
  ```
- Make sure you have also `pip install redis` in your galaxy Python virtual environment.
---

## TODO

- [x] <del>Handle pagination of records.</del> To properly handle this we need to support pagination at the plugin framework level. This is a bigger project that I would try in a separate PR. Limited to 100 records max for now.
- [x] Handle private/protected records
- [x] Make it work with the `export_remote` tool
  - [x] Provide detailed error messages when export fails due to invalid paths
- [x] Integrate with History export component
- [x] Integrate with History archival component
- [x] Write some tests
- [x] Refactor and clean up code
- [x] Remove temporal SSL verification ignore (for testing)
- [ ] [Optional] Import from non-public records. Requires user_context in fetch for authentication... possible?

---

## How to test the changes?

- [ ] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [ ] This is a refactoring of components with existing test coverage.
- [x] Instructions for manual testing are as follows:
  - Make sure you have all the requirements under [Manual Testing Requirements](#requirements)
  - See the [For Users](#for-users) section above for usage examples.

## License

- [x] I agree to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT).

davelopez committed:
Add basic Invenio file source plugin

Currently only basic support for listing records and downloading files in public records.

davelopez committed:
Add Invenio RDM plugin configuration sample

davelopez committed:
Explore record publishing

This has a severe limitation.
We can export a record with files (without much metadata) but we cannot reference it back directly form an export URI.
 The reason is how we construct the URI for regular file sources, which obviously are just file paths, while with DOI repositories, we need to differentiate between records and files only.

davelopez committed:
Use vault secret to store personal token

Also add example config to user_preferences_extra_conf.yml.sample

davelopez committed:
Add write intent option to File Sources

This options helps to identify when we are browsing a file source with the "intent to write" to it.

This is helpful to avoid listing those elements (directories/records) that might not be writable (protected, read-only, etc.) even if the file source itself is writable.

The plugin implementation should handle this option accordingly if necessary.

davelopez committed:
Add create_entry endpoint for File sources

This allows to create a new entry (directory/record) in those remote file sources
that supports it. By default, the endpoint will raise NotImplementedError unless
the plugin implements the _create_entry method.

davelopez committed:
Add RDMFilesSource subclass

Implement the _create_entry method for InvenioRDMFilesSource and some additional refactors.

davelopez committed:
Allow filtering only RDM plugins in the API

davelopez committed:
Update client API schema

davelopez committed:
Update FilesDialog services

- Add new API parameters
- Add new POST endpoint for creating entries
- Add some code docs

davelopez committed:
Allow to filter RDM FS in composable

davelopez committed:
Allow to filter RDM FS in FilesInput

davelopez committed:
Update FilesDialog

- Add new mode "source" to select only the root level of a file source.
- Allow filtering FS by RDM support
- Document props

davelopez committed:
Add ExportDOIForm component

This component allows to select or create new records in a DOI repository (RDM file source) for uploading exported artifacts (histories, datasets, etc.) to them.

davelopez committed:
Add option to export histories to DOI repository

davelopez committed:
Fix URL conversion between Invenio API / FS plugin API

davelopez committed:
Fix title display of record

v-localize will remove spaces between tags for some reason.

mvdbeek commented on lib/galaxy/files/sources/__init__.py:
```diff
@@ -53,7 +53,12 @@ class FilesSourceProperties(TypedDict):
+    """Options to control behavior of file source operations, such as realize_to, write_from and list."""
+
+    # Indicates access to the FS operation with intent to write.
+    # A file source can be "writeable" but, for example, some directories (or elements) may be restricted or read-only
```
I see, in that case this seems like something you'd apply when listing the contents ... I'd still call it writable though ?

davelopez replied:
Yeah, I wanted to differentiate it from the general `writeable` flag for the whole plugin. This `write_intent` makes only sense when the plugin is already "writeable" and provides just a way of signaling that "we are currently browsing the remote source to write a file on it, so please return only those directories that can be written to" as opposed from when we are browsing to import a file, which should list all o them, including read-only directories (or records).

Happy to change it to something more meaningful, I just think "writeable" could be confused with the general flag. Any other naming idea?

mvdbeek replied:
I don't think this has to be confusing, one is for the plugin, one is for the listing. They're not used in the same call, or are they ?

davelopez replied:
You are right, that is a good point. And besides, the API docs will clarify it anyway, I'll change it to `writeable` :+1:

mvdbeek commented on lib/galaxy/webapps/galaxy/api/remote_files.py:
```diff
@@ -63,6 +62,15 @@
+    title="Write Intent",
+    description=(
+        "Whether the query is made with the intention of writing to the source."
+        " If set to True, only entries that can be written to will be accessible."
```
```suggestion
        " If set to True, only entries that can be written to will be returned."
```
?

davelopez committed:
Fix wording in API doc

Co-authored-by: Marius van den Beek <m.vandenbeek@gmail.com>

davelopez committed:
Add TypedDicts to represent Directories and Files

+ Docstring

davelopez committed:
Add RDMRepositoryInteractor interface + refactor

davelopez committed:
Rename write_intent to writeable

davelopez committed:
Remove rdm_only parameter

davelopez committed:
Fix FilesSourcePluginList schema

The common field `browsable` was missing and only when browsable is True we require the `uri_root`.

We can also drop the `Extra.allow` config in FilesSourcePlugin model because the specific fields for different file sources are not returned in the API response since `plugins_to_dict` is called with for_serialization=False by default.

davelopez committed:
Add PluginKind to file sources and allow filtering

davelopez committed:
Do not publish records directly form Galaxy

It is safer to let the user decide when to publish the record. Galaxy should only deal with draft records with limited metadata.

davelopez committed:
Handle authorization in tool execution context

davelopez committed:
Validate source_path format on parse

davelopez committed:
Refactor FilesDialog unit tests

davelopez committed:
Fix typescript error in HistoryExport unit tests

davelopez committed:
Temporarily raise limit of records to 100

Until general pagination support is added for Remote File Sources.

davelopez committed:
Allow downloading files from draft and published records

davelopez committed:
Refactor Ok button disabled condition

davelopez committed:
Provide detailed error messages when invalid path

davelopez committed:
Improve prefix to avoid collisions

When exporting a history archive to a remote source.

davelopez committed:
Integrate on History Archive Export Selector

davelopez committed:
Rename DOI to RDM for consistency

davelopez committed:
Handle errors creating new records

davelopez committed:
Add external link to Invenio RDM

davelopez committed:
Add unit tests for ExportRDMForm component

davelopez committed:
Drop testing parameters to skip SSL verification

davelopez committed:
Refactor draft record check
